src/inference/hmc.jl

```diff
@@ -114,107 +109,146 @@ function sample(model::Model, alg::Hamiltonian;
               isa(alg, SGLD)  ? "SGLD"  :
               isa(alg, NUTS)  ? "NUTS"  : "Hamiltonian"
 
-    # Initialization
-    time_total = zero(Float64)
+    # TODO: figure out what does this line do
     n = reuse_spl_n > 0 ?
         reuse_spl_n :
         alg.n_iters
+
+    # Init samples
     samples = Array{Sample}(undef, n)
     weight = 1 / n
     for i = 1:n
         samples[i] = Sample(weight, Dict{Symbol, Any}())
     end
 
+    # Create VarInfo
     vi = if resume_from == nothing
         vi_ = VarInfo()
-        model(vi_, SampleFromUniform())
-        spl.info[:eval_num] += 1
+        runmodel!(model, vi_, SampleFromUniform())
         vi_
     else
         deepcopy(resume_from.info[:vi])
     end
 
+    if init_theta != nothing
+        println("Using init_theta=$init_theta")
+        init_theta_flat = foldl(vcat, init_theta)
+        theta_mask = map(x -> !ismissing(x), init_theta_flat)
+        theta = vi[spl]
+        theta[theta_mask] .= init_theta_flat[theta_mask]
+        vi[spl] = theta
+    end
+
+    # Convert to transformed sapce
     if spl.selector.tag == :default
         link!(vi, spl)
         runmodel!(model, vi, spl)
     end
 
-    # HMC steps
-    total_lf_num = 0
-    total_eval_num = 0
-    accept_his = Bool[]
-    PROGRESS[] && (spl.info[:progress] = ProgressMeter.Progress(n, 1, "[$alg_str] Sampling...", 0))
-    for i = 1:n
-        Turing.DEBUG && @debug "$alg_str stepping..."
+    # Init h, prop and adaptor
+    step(model, spl, vi, Val(true))
 
-        time_elapsed = @elapsed vi, is_accept = step(model, spl, vi, Val(i == 1))
-        time_total += time_elapsed
-
-        if is_accept # accepted => store the new predcits
+    # Sampling using AHMC
```

Comment:
Consider a bit factoring here to replace the branching statements with a dispatching based design, e.g.

```julia
step(a::AdaptiveHamiltonian) # or step(a::NUTS)
step(a:: HMC)
step(a:: HMCDA)
```

or

```julia
s::VarInfo, a::Alg, d::Adaptor = step(a::HMC, d::AbstractAdaptor) 
```

Reply:
Could do `step(model, spl::Sampler{H}, vi::VarInfo) where {H}` version